File: Binary_Search/Silver/P1_2017-Cow_Dance_Show/P1_2017-Cow_Dance_Show.cpp
// #include <bits/stdc++.h>

#include <iostream>
#include <fstream>
#include <algorithm>
#include <vector>
#include <queue>
#include <limits>

using namespace std;

int n, tMax;
vector<int> dances;

// k represents the size of the stage to support n amount of cows dancing
bool withinContract(int k);

int main() {
    // open input file
    ifstream fin("cowdance.in");
    // first line: n = number of cows, tMax = max stage time
    fin >> n >> tMax;
    // for next n lines, read in cow dance times
    for (int ln = 0; ln < n; ln++) {
        int t;
        fin >> t;
        dances.push_back(t);
    }
    fin.close();

    // Binary search for smallest possible stage size within contract time limit
    int lo = 1, hi = n;
    while (lo < hi) {
        int mid = lo + (hi - lo) / 2;
        if (!withinContract(mid))
            lo = mid + 1;   // increase stage size to reduce time on stage and be within contract time limit
        else
            hi = mid;
    }

    // write smallest possible stage size to output file
    ofstream pw("cowdance.out");
    pw << lo << endl;
    pw.close();
}

// Binary search - Heap/prority queue approach
// T: O(n * log n), M: O(n), where n is the number of cows
bool withinContract(int k) {
    // heapify first k cows
    priority_queue<int, vector<int>, greater<int>> stage(dances.begin(), dances.begin() + k);
    // elapse shortest dance time (root) to open dance spot and add next cow to stage
    int maxDance = numeric_limits<int>::min();
    for (int idx = k; idx < n; idx++) {
        int minDance = stage.top();
        stage.pop();
        stage.push(minDance + dances[idx]);
        maxDance = max(maxDance, minDance + dances[idx]);
    }
    // dance remaining cows on stage by searching max dance time
    return maxDance <= tMax;
}

/*
// Binary search - Array approach
// T: O(n^2 log n), M: O(n), where n is the number of cows
bool within_contract(int k) {
    vector<int> stage(dances.begin(), dances.begin() + k);
    int t_cnt = 0, d_idx = k;
    while (d_idx < n) {
        // find the shortest dance time on stage
        int minDance = *min_element(stage.begin(), stage.end());
        t_cnt += minDance;
        if (t_cnt > tMax)
            return false;
        // elapse time by shorest dance time to open dance spot and add new cow to stage
        for (int i = 0; i < k; i++) {
            stage[i] -= minDance;
            if (stage[i] == 0) {
                stage[i] = dances[d_idx];
                d_idx++;
            }
        }
    }
    return t_cnt + *max_element(stage.begin(), stage.end()) <= tMax;
}
*/

File: Binary_Search/Silver/P1_2017-Cow_Dance_Show/P1_2017-Cow_Dance_Show.py
# Binary Search - Heap approach
# T: O(n * log n), M: O(n), where n is number of cows

import heapq

# open input file
with open("cowdance.in") as r:
	# first line n = cows, t_max = maximum time for all cows to dance
	n, t_max = map(int, r.readline().split())
	# for the next n lines read each cow's dance time
	dances = [int(r.readline()) for ln in range(n)]

# k represents the size of the stage to support n amount of cows dancing
def within_contract(k):
	stage = dances[:k]
	heapq.heapify(stage)
	# elapse (heappop) shortest dance time and add next cow to dance
	for idx in range(k, n):
		min_dance = heapq.heappop(stage)
		heapq.heappush(stage, min_dance + dances[idx])
	return max(stage) <= t_max

# Search for smallest possible stage size within contract limit using binary search
lo, hi = 1, n
while lo < hi:
	mid = (lo + hi) // 2
	if not within_contract(k = mid):
		lo = mid + 1    # increase stage size to reduce time on stage and be within contract time limit
	else:
		hi = mid

# Write searched item into output file
print(hi, file=open("cowdance.out", "w"))

'''
# Binary Search - Array approach
# T: O(n^2 * log n), M: O(n), where n is number of cows

# open input file
with open("cowdance.in") as r:
	# first line n = cows, t_max = maximum time for all cows to dance
	n, t_max = map(int, r.readline().split())
	# for the next n lines read each cow's dance time
	dances = [int(r.readline()) for ln in range(n)]

# k represents the size of the stage to support n amount of cows dancing
def within_contract(k):
	stage = dances[:k]
	t_cnt = 0
	d_idx = k
	while d_idx < n:
		min_dance = min(stage)
		t_cnt += min_dance
		if t_cnt > t_max:
			return False
		# elapse time by shortest dance time to open dance spot and add next cow to dance
		for s_idx, s_time in enumerate(stage):
			stage[s_idx] -= min_dance
			if stage[s_idx] == 0:
				stage[s_idx] = dances[d_idx]
				d_idx += 1
	# let cows dance remaining time
	return t_cnt + max(stage) <= t_max

# Search for smallest possible stage size within contract limit using binary search
lo, hi = 1, n
while lo < hi:
	mid = (lo + hi) // 2
	if not within_contract(k = mid):
		lo = mid + 1    # increase stage size to reduce time on stage and be within contract time limit
	else:
		hi = mid
# Write searched item into output file
print(hi, file=open("cowdance.out", "w"))
'''
